src/acl.c

```diff
@@ -937,6 +963,71 @@ static void ACLAddAllowedFirstArg(aclSelector *selector, unsigned long id, const
     selector->allowed_firstargs[id][items - 1] = NULL;
 }
 
+/* Clean up and return error from database permission operations. */
+static int ACLDatabasePermissionError(intset *new_dbs, sds dblist, sds *tokens, int count, int error_code) {
+    if (tokens) sdsfreesplitres(tokens, count);
+    if (new_dbs) intsetFree(new_dbs);
+    if (dblist) sdsfree(dblist);
+    errno = error_code;
+    return C_ERR;
+}
+
+/* Helper function to add database permissions to a selector. */
+static int ACLSetSelectorDatabasePermissions(aclSelector *selector, const char *dbs_str) {
+    /* Initialize the database intset */
+    intset *new_dbs = intsetNew();
+
+    if (selector->flags & SELECTOR_FLAG_ALLDBS) {
+        selector->flags &= ~SELECTOR_FLAG_ALLDBS;
+    }
+
+    sds dblist = sdsnew(dbs_str);
+    size_t len = sdslen(dblist);
+
+    /* Reject empty list, leading or trailing commas */
+    if (len == 0 || dblist[0] == ',' || dblist[len - 1] == ',') {
+        return ACLDatabasePermissionError(new_dbs, dblist, NULL, 0, EINVAL);
+    }
+
+    int count;
+    sds *tokens = sdssplitlen(dblist, len, ",", 1, &count);
+
+    for (int i = 0; i < count; i++) {
+        /* Reject empty tokens */
+        if (sdslen(tokens[i]) == 0) {
+            return ACLDatabasePermissionError(new_dbs, dblist, tokens, count, EINVAL);
+        }
+
+        char *endptr = NULL;
+        errno = 0;
+        int64_t dbid = strtoll(tokens[i], &endptr, 10);
+
+        /* Reject invalid input format */
+        if (*endptr != '\0') {
+            return ACLDatabasePermissionError(new_dbs, dblist, tokens, count, EINVAL);
+        }
+
+        /* Reject out of range values */
+        if (errno == ERANGE || dbid < 0 || dbid >= server.dbnum) {
```

Comment:
```suggestion
        if (errno == ERANGE || dbid < 0) {
```
I'm wondering if we should be verifying the validity of the db number on the current instance. I can imagine someone has a standardized set of ACLs, and they want to apply it widely even if all servers don't have everything configured. In AWS we decouple the configuration of databases from the ACL configuration, and we would need to add extra code to make sure they are consistent. 

We've also talked about making the number of DBs dynamic as well, so if we do we just need to make sure we check that no users have high DBs.

Reply:
I think we should have this check because it would prevent a lot of misuse (and also make it faster to search smaller intset compared to intset that has a lot of `dbid >= server.dbnum`).
For the case of dynamic number of DBs, I would assume server administrator has to keep ACLs in sync with current configuration (at Yandex Cloud we also have that logic decoupled but at any given moment for one node there is only one valid server.dbnum, so it shouldn't be that hard to validate, WDYT?)

Reply:
I agree with @dvkashapov to keep it bounded as we will always have a upper bound.